nebu/bakedpdf.py

```diff
@@ -0,0 +1,129 @@
+from lxml import etree
+
+from cnxepub.models import Binder, TranslucentBinder, Document
+from cnxepub.formatters import SingleHTMLFormatter
+
+
+__all__ = (
+    'assembler',
+)
+
+
+"""`Document`s require that modules contain a body tag, but this assembler
+allows for building a single/collection HTML with empty modules, therefore
+we must use this variable in place of data when building `Document` models.
+When updating this also update the tests.
+"""
+DATA_WHEN_MODULE_NOT_FOUND = "<body>Module with id {} goes here.</body>"
+
+
+def assembler(source_dir, out_dir='.', out_fname='collection'):
+    """Assembles a single HTML file from a collection of module HTML files, a
+    CollXML file defines their ordering/structure.
+    """
+    out_loc = out_dir / (out_fname + '.xhtml')
+    collxml = (source_dir / 'collection.xml').open('rb')
+
+    def main():
+        binder = collxml_to_binder(collxml)
+
+        with out_loc.open('w') as f:
+            f.write(str(SingleHTMLFormatter(binder)))
+
+    def collxml_to_binder(collxml):
+        tags = [
+            '{http://cnx.rice.edu/collxml}subcollection',
+            '{http://cnx.rice.edu/collxml}collection',
+            '{http://cnx.rice.edu/collxml}module',
+            '{http://cnx.rice.edu/mdml}title',
+        ]
+
+        collxml_etree = etree.iterparse(collxml, events=('start', 'end'),
+                                        tag=tags, remove_blank_text=True)
+
+        root_adapter = build_adapters_tree(collxml_etree)
+
+        binder = root_adapter.to_model()
+        return binder
+
+    def build_adapters_tree(collxml_etree):
+        """Binders require bottom-up construction (sub-binders are passed into
+        the constructor). So we'll (1) parse the element tree top-down and
+        store the needed information in Adapter nodes and then (2) use
+        post-order traversalon the Adapter tree to construct Binders from their
+        sub-binders.
+        """
```

Comment:
Probably, and I think that would be more like Karen's implementation, but still we decided to go for my implementation because it's clearer.

Reply:
The reason why I tried implementing it differently is exactly because I don't like event based parsing as it's hard to understand. I think this is ok if it works.

Reply:
Hm... Ok. I'll rephrase: I guess it was my call to choose which implementation and I chose mine because **Separation of Concerns**.